Prepare for GCP IPAM

tgraf opened the pull request.

errordeveloper commented on operator/main.go:
```diff
@@ -256,7 +264,7 @@ func runOperator(cmd *cobra.Command) {
 
 		if option.Config.SyncK8sNodes {
-			if err := runNodeWatcher(); err != nil {
+			if err := runNodeWatcher(nodeManager); err != nil {
```
Does this function have any effect if `nodeManager` wasn't initialised, i.e. neither of `option.Config.IPAM` cases were applied?

tgraf replied:
Yes, the `runNodeWatcher()` and dependent code will check if `nodeManager` is non-nil and only notify the node manager about modifications if so.

errordeveloper replied:
Ok, so does it need to be a separate thing, or it could run as one of the tasks inside `nodeManager`?

tgraf replied:
The Kubernetes node watcher is also used for other things besides the IPAM hence it is separate right now.

errordeveloper commented on operator/azure.go:
```diff
@@ -56,12 +56,12 @@ func startAzureAllocator(clientQPSLimit float64, clientBurst int) error {
+		return nil, fmt.Errorf("unable to initialize Azure node manager: %s", err)
 	}
 
 	instances.Resync(context.TODO())
```
Do we want `Resync` and the goroutine below to remain under `operator` (either in main package or a sub-package), or it would be fine to move somewhere into `pkg/`? I can imagine it maybe a convention where it's desirable for anything that uses log package and starts go routines to be out of `pkg/`, but no idea if that applies or I'm overthinking it.

E.g. there `pkg/azure/ipam` and  `pkg/ipam/allocator`, but for some reason this function lives here, so does the AWS conterpart. Would it better to move these into `pkg/` or something like `operator/tasks/{azure,aws}_allocator`?

tgraf replied:
Good point, the code is basically the same for all IPAM modes at this point. I'll move it into the generic ipam package.

tgraf replied:
Done

tgraf committed:
ipam: Move iterator logic into generic InstanceMap

Move code to iterate over interfaces and addresses into the InstanceMap.
This allows to defer deep copying until it is really necessary.

This also allows to move the logic to reserve IPs of all interfaces of
all instances into the generic GroupPoolAllocator for reuse in future
IPAM code.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Simplify locking in IPAM node interactions

Don't hold any locks when calling into IPAM specific operations. Use
node.mutex exclusively to protect data access and keep locking periods
as short as possible.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add InstanceMap.GetInterface

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Don't recalculate if resource is not attached yet

There is no point in attempting recalculation of required addresses
unless the resource has been attached.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Move instance ID retrieval into generic code

This removes the need to track the CiliumNode resource in the Azure
specific code entirely.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Fix data race in node manager test

Fixes:
```
WARNING: DATA RACE
Read at 0x00c0001b48c0 by goroutine 37:
  github.com/cilium/cilium/pkg/ipam.(*nodeOperationsMock).PrepareIPAllocation()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node_manager_test.go:103 +0x1c5
  github.com/cilium/cilium/pkg/ipam.(*Node).determineMaintenanceAction()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node.go:492 +0x184
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node.go:528 +0x53
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node.go:613 +0x69
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func1()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:205 +0x8b
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:206 +0x4b9

Previous write at 0x00c0001b48c0 by goroutine 35:
  github.com/cilium/cilium/pkg/ipam.(*nodeOperationsMock).AllocateIPs()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node_manager_test.go:111 +0x117
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node.go:564 +0xa0d
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node.go:613 +0x69
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func1()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:205 +0x8b
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/vagrant/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:206 +0x4b9
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam/allocator: Allow allocating IPs without knowing pool ID

Ideally, the pool ID is known. If not known, the list of pools can be
searched to identify the correct pool.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Remove global nodeManager variable

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Move initial sync and background sync in generic code

Signed-off-by: Thomas Graf <thomas@cilium.io>
